File: tools/clients/generate.py
#!/usr/bin/env python3
"""Generate only owned files, in a temporary directory; --check never writes them."""
import argparse
import hashlib
import json
import os
from pathlib import Path
import shutil
import subprocess
import tempfile

ROOT = Path(__file__).resolve().parents[2]
CONFIG = json.loads((ROOT / 'tools/clients/config.json').read_text())

def run(*args):
    subprocess.run(args, cwd=ROOT, check=True)

def dump(path, value):
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(value, ensure_ascii=False, indent=2) + '\n')

def command_reference(spec, operations):
    lines = ['# API command reference (generated)', '',
             'Find API commands by resource. Use `wavectl <resource> <command> --help` for installed flags,',
             '`wavectl schema <resource> <command> --request` for request fields, or `--example` for a template.',
             'Use `--dry-run` to validate and preview a request. For workflow commands, follow the task and tool references linked from SKILL.md.', '']
    groups = {}
    for name, op in operations.items():
        groups.setdefault(op['command'].split()[0], []).append((name, op))
    for group, entries in sorted(groups.items()):
        lines.extend([f'## {group}', '', '| Command | Required path flags | Required body fields | Description |', '| --- | --- | --- | --- |'])
        for name, op in entries:
            paths = ', '.join('`--' + p['name'].replace('_', '-').lower() + '`' for p in op['parameters'] if p['in'] == 'path') or '—'
            request = (op.get('requestBody') or {}).get('content', {}).get('application/json', {}).get('schema', {})
            if '$ref' in request:
                request = spec['components']['schemas'][request['$ref'].split('/')[-1]]
            body = ', '.join('`--' + key.replace('_', '-') + '`' for key in request.get('required', [])) or '—'
            if op.get('transport') == 'upload': body = '`--file`'
            if op.get('transport') == 'download': body = '`--output`'
            lines.append(f"| `wavectl {op['command']}` | {paths} | {body} | {op['summary'].replace('|', '/')} |")
        lines.append('')
    lines += ['Path IDs also accept positional arguments in the order shown by help; session/task IDs support `--session` / `--task` aliases.',
              'Choose field flags or full `--body @file.json`; do not combine them.',
              'Paginated commands support `--all --format jsonl`; uploads support `--file - --filename NAME`.', '']
    return '\n'.join(lines)

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--check', action='store_true')
    args = parser.parse_args()
    with tempfile.TemporaryDirectory(prefix='wave-generate-') as directory:
        stage = Path(directory)
        stage.chmod(0o755)
        specfile = stage / 'api/openapi.json'
        specfile.parent.mkdir()
        run('node', 'tools/clients/convert.cjs', 'internal/platform/apidocs/swagger.json', str(specfile))
        chinese = stage / 'api/openapi.zh-CN.json'
        run('node', 'tools/clients/convert.cjs', 'internal/platform/apidocs/swagger.json', str(chinese), 'zh-CN')
        spec = json.loads(specfile.read_text())
        semantics = json.loads((ROOT / 'tools/clients/operations.json').read_text())
        operations = {}
        for path, methods in spec['paths'].items():
            for method, op in methods.items():
                name = op['operationId']
                if name in operations:
                    raise ValueError(f'duplicate operationId: {name}')
                operations[name] = dict(semantics[name], id=name, method=method.upper(), path=path,
                    summary=op.get('summary', ''), parameters=op.get('parameters', []),
                    requestBody=op.get('requestBody'), responses=op['responses'],
                    authenticated=bool(op.get('security', spec.get('security', []))))
        if operations.keys() != semantics.keys():
            raise ValueError('stale operation mapping: ' + str(semantics.keys() - operations.keys()))
        commands = [v['command'] for v in operations.values()]
        if len(commands) != len(set(commands)):
            raise ValueError('duplicate CLI command')
        docker = ['docker', 'run', '--rm', '--network', 'none', '--user', f'{os.getuid()}:{os.getgid()}',
                  '-v', f'{stage}:/local', CONFIG['generatorImage']]
        run(*docker, 'validate', '-i', '/local/api/openapi.json')
        run(*docker, 'validate', '-i', '/local/api/openapi.zh-CN.json')
        targets = {
            'go': {'packageName': 'generated', 'isGoSubmodule': True},
            'python': {'packageName': 'wave_ai_generated', 'packageVersion': CONFIG['version']},
            'typescript-fetch': {'npmName': CONFIG['npmPackage'], 'npmVersion': CONFIG['version']},
        }
        for target, settings in targets.items():
            settings.update(hideGenerationTimestamp=True, disallowAdditionalPropertiesIfNotPresent=False)
            configfile = stage / f'{target}.json'
            dump(configfile, settings)
            run(*docker, 'generate', '-i', '/local/api/openapi.json', '-g', target,
                '-o', f'/local/raw/{target}', '-c', f'/local/{target}.json',
                '--global-property', 'apiTests=false,modelTests=false,apiDocs=false,modelDocs=false')
        go = stage / 'sdks/go/generated'
        go.mkdir(parents=True)
        for source in (stage / 'raw/go').glob('*.go'):
            shutil.copy(source, go / source.name)
        shutil.copytree(stage / 'raw/python/wave_ai_generated', stage / 'sdks/python/wave_ai_generated')
        shutil.copytree(stage / 'raw/typescript-fetch/src', stage / 'sdks/typescript/src/generated')
        # Node ESM requires explicit extensions. This mechanical normalization
        # is deterministic and does not change generated API behavior.
        import re
        for file in (stage / 'sdks/typescript/src/generated').rglob('*.ts'):
            file.write_text(re.sub(r"(from\s+['\"])(\.[^'\"]+)(['\"])",
                lambda m: m[1] + m[2] + ('' if m[2].endswith('.js') else '.js') + m[3], file.read_text()))
        shutil.copy(specfile, stage / 'sdks/go/openapi.json')
        for destination in ['sdks/go/operations.json', 'sdks/python/wave_ai/operations.json']:
            dump(stage / destination, operations)
        ts = stage / 'sdks/typescript/src/operations.ts'
        ts.write_text('// Generated by tools/clients/generate.py. Do not edit.\nexport const operations = ' +
                      json.dumps(operations, ensure_ascii=False, indent=2) + ' as const;\n' +
                      'export type OperationID = keyof typeof operations;\n')
        reference = stage / 'skills/wave-client/references/commands.md'
        reference.parent.mkdir(parents=True)
        reference.write_text(command_reference(spec, operations))
        dump(stage / 'tools/clients/manifest.json', dict(version=CONFIG['version'],
            sourceSHA256=hashlib.sha256((ROOT / 'internal/platform/apidocs/swagger.json').read_bytes()).hexdigest(),
            openapiSHA256=hashlib.sha256(specfile.read_bytes()).hexdigest(),
            openapiChineseSHA256=hashlib.sha256(chinese.read_bytes()).hexdigest(),
            defaultLanguage='en',
            generatorImage=CONFIG['generatorImage'], operations=len(operations)))
        owned = ['api/openapi.json', 'api/openapi.zh-CN.json', 'sdks/go/generated', 'sdks/go/operations.json', 'sdks/go/openapi.json',
                 'sdks/python/wave_ai_generated', 'sdks/python/wave_ai/operations.json',
                 'sdks/typescript/src/generated', 'sdks/typescript/src/operations.ts',
                 'skills/wave-client/references/commands.md', 'tools/clients/manifest.json']
        changed = []
        for name in owned:
            source, dest = stage / name, ROOT / name
            def files(p):
                if not p.exists(): return {}
                if p.is_file(): return {'': p.read_bytes()}
                return {str(f.relative_to(p)): f.read_bytes() for f in p.rglob('*') if f.is_file()
                        and '__pycache__' not in f.parts}
            if files(source) != files(dest):
                changed.append(name)
                if not args.check:
                    if dest.is_dir(): shutil.rmtree(dest)
                    dest.parent.mkdir(parents=True, exist_ok=True)
                    if source.is_dir(): shutil.copytree(source, dest)
                    else: shutil.copy(source, dest)
        if args.check and changed:
            raise SystemExit('out of date: ' + ', '.join(changed))
        print(f"{'Checked' if args.check else 'Generated'} {len(operations)} operations across three SDKs")

if __name__ == '__main__':
    main()

File: tools/clients/generate.go
package main

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

var generatedPaths = []string{"api/openapi.json", "api/openapi.zh-CN.json", "sdks/go/generated", "sdks/go/operations.json", "sdks/go/openapi.json", "sdks/python/wave_ai_generated", "sdks/python/wave_ai/operations.json", "sdks/typescript/src/generated", "sdks/typescript/src/operations.ts", "skills/wave-client/references/commands.md", "tools/clients/manifest.json"}

func operationMap(spec, semantics object) (object, []string, error) {
	operations := object{}
	var order []string
	commands := map[string]bool{}
	for _, path := range sortedKeys(obj(spec["paths"])) {
		methods := obj(at(spec, "paths", path))
		for _, method := range sortedKeys(methods) {
			op := obj(methods[method])
			id := str(op["operationId"])
			if id == "" {
				return nil, nil, fmt.Errorf("missing operationId: %s %s", method, path)
			}
			if _, exists := operations[id]; exists {
				return nil, nil, fmt.Errorf("duplicate operationId: %s", id)
			}
			metadata, exists := semantics[id]
			if !exists {
				return nil, nil, fmt.Errorf("missing operation metadata: %s", id)
			}
			item := object{}
			for key, value := range obj(metadata) {
				item[key] = value
			}
			command := str(item["command"])
			if command == "" || commands[command] {
				return nil, nil, fmt.Errorf("missing/duplicate CLI command for %s", id)
			}
			commands[command] = true
			item["id"] = id
			item["method"] = strings.ToUpper(method)
			item["path"] = path
			item["summary"] = op["summary"]
			params := op["parameters"]
			if params == nil {
				params = []any{}
			}
			item["parameters"] = params
			item["requestBody"] = op["requestBody"]
			item["responses"] = op["responses"]
			security, exists := op["security"]
			if !exists {
				security = spec["security"]
			}
			item["authenticated"] = len(list(security)) > 0
			operations[id] = item
			order = append(order, id)
		}
	}
	for id := range semantics {
		if _, exists := operations[id]; !exists {
			return nil, nil, fmt.Errorf("stale operation metadata: %s", id)
		}
	}
	return operations, order, nil
}

func commandReference(spec, operations object, order []string) string {
	var b strings.Builder
	b.WriteString("# API command reference (generated)\n\nFind API commands by resource. Use `wavectl <resource> <command> --help` for installed flags,\n`wavectl schema <resource> <command> --request` for request fields, or `--example` for a template.\nUse `--dry-run` to validate and preview a request. For workflow commands, follow the task and tool references linked from SKILL.md.\n\n")
	groups := map[string][]object{}
	for _, id := range order {
		op := obj(operations[id])
		group := strings.Fields(str(op["command"]))[0]
		groups[group] = append(groups[group], op)
	}
	flags := func(values []string) string {
		if len(values) == 0 {
			return "—"
		}
		return strings.Join(values, ", ")
	}
	flag := func(name string) string { return "`--" + strings.ToLower(strings.ReplaceAll(name, "_", "-")) + "`" }
	for _, group := range sortedKeys(groups) {
		fmt.Fprintf(&b, "## %s\n\n| Command | Required path flags | Required body fields | Description |\n| --- | --- | --- | --- |\n", group)
		for _, op := range groups[group] {
			var paths, body []string
			for _, p := range list(op["parameters"]) {
				if at(p, "in") == "path" {
					paths = append(paths, flag(str(at(p, "name"))))
				}
			}
			request := obj(at(op, "requestBody", "content", "application/json", "schema"))
			if ref := str(request["$ref"]); ref != "" {
				request = obj(at(spec, "components", "schemas", strings.TrimPrefix(ref, "#/components/schemas/")))
			}
			for _, name := range list(request["required"]) {
				body = append(body, flag(str(name)))
			}
			switch op["transport"] {
			case "upload":
				body = []string{"`--file`"}
			case "download":
				body = []string{"`--output`"}
			}
			fmt.Fprintf(&b, "| `wavectl %s` | %s | %s | %s |\n", op["command"], flags(paths), flags(body), strings.ReplaceAll(str(op["summary"]), "|", "/"))
		}
		b.WriteString("\n")
	}
	b.WriteString("Path IDs also accept positional arguments in the order shown by help; session/task IDs support `--session` / `--task` aliases.\nChoose field flags or full `--body @file.json`; do not combine them.\nPaginated commands support `--all --format jsonl`; uploads support `--file - --filename NAME`.\n")
	return b.String()
}

func (t *tool) generate(ctx context.Context, checkOnly bool) error {
	stage, err := os.MkdirTemp("", "wave-generate-*")
	if err != nil {
		return err
	}
	defer os.RemoveAll(stage)
	if err = os.Chmod(stage, 0755); err != nil {
		return err
	}
	if err = os.MkdirAll(filepath.Join(stage, "api"), 0755); err != nil {
		return err
	}
	for _, locale := range []struct{ file, language string }{{"openapi.json", "en"}, {"openapi.zh-CN.json", "zh-CN"}} {
		if err = t.run(ctx, t.root, nil, "node", "tools/clients/convert.cjs", "internal/platform/apidocs/swagger.json", filepath.Join(stage, "api", locale.file), locale.language); err != nil {
			return err
		}
	}
	var spec, semantics object
	if err = readJSON(filepath.Join(stage, "api/openapi.json"), &spec); err != nil {
		return err
	}
	if err = readJSON(t.path("tools/clients/operations.json"), &semantics); err != nil {
		return err
	}
	operations, order, err := operationMap(spec, semantics)
	if err != nil {
		return err
	}
	if err = t.generateSDKs(ctx, stage); err != nil {
		return err
	}
	if err = copyFile(filepath.Join(stage, "api/openapi.json"), filepath.Join(stage, "sdks/go/openapi.json")); err != nil {
		return err
	}
	for _, dest := range []string{"sdks/go/operations.json", "sdks/python/wave_ai/operations.json"} {
		if err = writeJSON(filepath.Join(stage, dest), operations); err != nil {
			return err
		}
	}
	data, err := jsonBytes(operations)
	if err != nil {
		return err
	}
	ts := "// Generated by go run ./tools/clients generate. Do not edit.\nexport const operations = " + strings.TrimSuffix(string(data), "\n") + " as const;\nexport type OperationID = keyof typeof operations;\n"
	if err = writeFile(filepath.Join(stage, "sdks/typescript/src/operations.ts"), []byte(ts), 0644); err != nil {
		return err
	}
	if err = writeFile(filepath.Join(stage, "skills/wave-client/references/commands.md"), []byte(commandReference(spec, operations, order)), 0644); err != nil {
		return err
	}
	manifest := object{"version": t.config.Version, "defaultLanguage": "en", "generatorImage": t.config.GeneratorImage, "operations": len(operations)}
	for key, path := range map[string]string{"sourceSHA256": t.path("internal/platform/apidocs/swagger.json"), "openapiSHA256": filepath.Join(stage, "api/openapi.json"), "openapiChineseSHA256": filepath.Join(stage, "api/openapi.zh-CN.json")} {
		sum, err := hashFile(path)
		if err != nil {
			return err
		}
		manifest[key] = sum
	}
	if err = writeJSON(filepath.Join(stage, "tools/clients/manifest.json"), manifest); err != nil {
		return err
	}
	if err = ctx.Err(); err != nil {
		return err
	}
	if err = t.publishGenerated(stage, checkOnly); err != nil {
		return err
	}
	fmt.Fprintf(t.out, "%s %d operations across three SDKs\n", map[bool]string{true: "Checked", false: "Generated"}[checkOnly], len(operations))
	return nil
}

func (t *tool) generateSDKs(ctx context.Context, stage string) error {
	docker := []string{"docker", "run", "--rm", "--network", "none"}
	if os.Getuid() >= 0 {
		docker = append(docker, "--user", fmt.Sprintf("%d:%d", os.Getuid(), os.Getgid()))
	}
	docker = append(docker, "-v", stage+":/local", t.config.GeneratorImage)
	run := func(args ...string) error {
		return t.run(ctx, t.root, nil, append(append([]string{}, docker...), args...)...)
	}
	for _, file := range []string{"openapi.json", "openapi.zh-CN.json"} {
		if err := run("validate", "-i", "/local/api/"+file); err != nil {
			return err
		}
	}
	targets := []struct {
		name     string
		settings object
	}{{"go", object{"packageName": "generated", "isGoSubmodule": true}}, {"python", object{"packageName": "wave_ai_generated", "packageVersion": t.config.Version}}, {"typescript-fetch", object{"npmName": t.config.NPMPackage, "npmVersion": t.config.Version}}}
	for _, target := range targets {
		target.settings["hideGenerationTimestamp"] = true
		target.settings["disallowAdditionalPropertiesIfNotPresent"] = false
		if err := writeJSON(filepath.Join(stage, target.name+".json"), target.settings); err != nil {
			return err
		}
		if err := run("generate", "-i", "/local/api/openapi.json", "-g", target.name, "-o", "/local/raw/"+target.name, "-c", "/local/"+target.name+".json", "--global-property", "apiTests=false,modelTests=false,apiDocs=false,modelDocs=false"); err != nil {
			return err
		}
	}
	files, err := filepath.Glob(filepath.Join(stage, "raw/go/*.go"))
	if err != nil {
		return err
	}
	for _, name := range files {
		if err = copyFile(name, filepath.Join(stage, "sdks/go/generated", filepath.Base(name))); err != nil {
			return err
		}
	}
	if err = copyTree(filepath.Join(stage, "raw/python/wave_ai_generated"), filepath.Join(stage, "sdks/python/wave_ai_generated")); err != nil {
		return err
	}
	ts := filepath.Join(stage, "sdks/typescript/src/generated")
	if err = copyTree(filepath.Join(stage, "raw/typescript-fetch/src"), ts); err != nil {
		return err
	}
	files, err = treeFiles(ts)
	if err != nil {
		return err
	}
	imports := regexp.MustCompile(`(from\s+['"])(\.[^'"]+)(['"])`)
	for _, file := range files {
		if filepath.Ext(file) != ".ts" {
			continue
		}
		data, err := os.ReadFile(file)
		if err != nil {
			return err
		}
		data = imports.ReplaceAllFunc(data, func(match []byte) []byte {
			parts := imports.FindSubmatch(match)
			if strings.HasSuffix(string(parts[2]), ".js") {
				return match
			}
			return []byte(string(parts[1]) + string(parts[2]) + ".js" + string(parts[3]))
		})
		if err = writeFile(file, data, 0644); err != nil {
			return err
		}
	}
	return nil
}

func (t *tool) publishGenerated(stage string, checkOnly bool) error {
	var changed []string
	for _, name := range generatedPaths {
		if _, err := os.Stat(filepath.Join(stage, name)); err != nil {
			return fmt.Errorf("incomplete generation: %w", err)
		}
		same, err := samePath(filepath.Join(stage, name), t.path(name))
		if err != nil {
			return err
		}
		if !same {
			changed = append(changed, name)
		}
	}
	if checkOnly && len(changed) > 0 {
		return fmt.Errorf("out of date: %s; run make clients-generate", strings.Join(changed, ", "))
	}
	if !checkOnly {
		for _, name := range changed {
			if err := replacePath(filepath.Join(stage, name), t.path(name)); err != nil {
				return err
			}
		}
	}
	return nil
}
